org.eclipse.wildwebdeveloper.xml/src/org/eclipse/wildwebdeveloper/xml/internal/lsp/ClosingTagReconcilingStrategy.java

```diff
@@ -0,0 +1,115 @@
+package org.eclipse.wildwebdeveloper.xml.internal.lsp;
+
+import java.net.URI;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import org.eclipse.jface.text.BadLocationException;
+import org.eclipse.jface.text.DocumentEvent;
+import org.eclipse.jface.text.IDocument;
+import org.eclipse.jface.text.IRegion;
+import org.eclipse.jface.text.ITextViewer;
+import org.eclipse.jface.text.reconciler.DirtyRegion;
+import org.eclipse.jface.text.reconciler.IReconcilingStrategy;
+import org.eclipse.lsp4e.LSPEclipseUtils;
+import org.eclipse.lsp4e.LanguageServiceAccessor;
+import org.eclipse.lsp4e.LanguageServiceAccessor.LSPDocumentInfo;
+import org.eclipse.lsp4j.TextDocumentIdentifier;
+import org.eclipse.lsp4j.TextDocumentPositionParams;
+import org.eclipse.swt.widgets.Display;
+import org.eclipse.wildwebdeveloper.xml.internal.XMLLanguageServerAPI;
+
+public class ClosingTagReconcilingStrategy implements IReconcilingStrategy {
+
+
+	private ITextViewer viewer;
+
+	public void documentChanged(DocumentEvent e) {
+		update(e);	
+	}
+
+	public void update(DocumentEvent event) {
+		if (event == null || viewer == null) {
+			return;
+		}
+		IDocument document = event.getDocument();
+		if (document == null || event == null || event.getLength() != 0 || event.getText().length() != 1) {
+			return;
+		}
+
+		int offset = event.getOffset() + 1;
+		char c = event.getText().charAt(0);
+		if (c != '>' && c != '/') {
+			return;
+		}
+		URI uri = LSPEclipseUtils.toUri(document);
+		if (uri == null) {
+			return;
+		}
+
+		TextDocumentIdentifier identifier = new TextDocumentIdentifier(uri.toString());
+		Optional<LSPDocumentInfo> info = LanguageServiceAccessor
+				.getLSPDocumentInfosFor(document, (capabilities) -> true).stream()
+				.filter(doc -> (doc.getLanguageClient() instanceof XMLLanguageServerAPI)).findAny();
+		if (!info.isEmpty()) {
+			// The document is bound with XML language server, consumes the xml/closeTag
+			final Display display = viewer.getTextWidget().getShell().getDisplay();
+			CompletableFuture.supplyAsync(() -> {
+				try {
+					// Wait for textDocument/didChange
+					Thread.sleep(100);
+				} catch (InterruptedException ex) {
+				}
+				try {
+					TextDocumentPositionParams params = LSPEclipseUtils.toTextDocumentPosistionParams(uri, offset,
+							document);
+					// consumes xml/closeTag from XML language server
+					((XMLLanguageServerAPI) info.get().getLanguageClient()).closeTag(params).thenAccept(r -> {
+						if (r != null) {
+							display.asyncExec(() -> {
+								try {
+									// we receive a text like 
+									// $0</foo>
+									// $0 should be used for set the cursor.
+									String text = r.snippet.replace("$0", "");

```

Comment:
One days, when snippets are usable more widely in LSP, we'll extract the LSCompletionProposal stuff that supports snippets as an API so such extension will be able to support it.
Or maybe the snippets will be generalized in LSP first and we'll be able to get rid of this custom operation before we have need for an API in LSP4E!